examples/tree/plot_cost_complexity_pruning.py

```diff
@@ -0,0 +1,100 @@
+# Minimal cost complexity pruning recursively finds the node with the
+# "weakest link". The weakest link is characterized by an effective alpha,
+# where the nodes with the smallest effective alpha are pruned first.
+# scikit-learn provides a
+# :func:`DecisionTreeClassifier.cost_complexity_pruning_path`
+# that returns the effective alphas and the corresponding total leaf impurities
+# at each step of the pruning process. As alpha increases, more
+# of the tree is pruned, which increases the total impurity of its leaves.
+# In the following plot, the maximum effective alpha value is removed,
+# because it is the trivial tree with only one node.
+X, y = load_breast_cancer(return_X_y=True)
+X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
+
+clf = DecisionTreeClassifier(random_state=0)
+path = clf.cost_complexity_pruning_path(X_train, y_train)
+ccp_alphas, impurities = path.ccp_alphas, path.impurities
+fig, ax = plt.subplots()
+ax.plot(ccp_alphas[:-1], impurities[:-1], drawstyle="steps-post")
+ax.set_xlabel("effective alpha")
+ax.set_ylabel("total impurity of leaves")
+ax.set_title("Total Impurity vs effective alpha for training set")
+plt.show()
+
+###############################################################################
+# Next, we train a decision tree using the effective alphas. The last value
+# in ``ccp_alphas`` is the alpha value that prunes the whole tree,
+# leaving the tree, ``clfs[-1]``, with one node.
+clfs = []
+for ccp_alpha in ccp_alphas:
+    clf = DecisionTreeClassifier(random_state=0, ccp_alpha=ccp_alpha)
+    clf.fit(X_train, y_train)
+    clfs.append(clf)
+print("Number of nodes in the last tree is: {} with ccp_alpha: {}".format(
+      clfs[-1].tree_.node_count, ccp_alphas[-1]))
+
+###############################################################################
+# For the remainder of this example, we remove the last element in
+# ``clfs`` and ``ccp_alphas``, because it is the trivial tree with only one
+# node. Here we show that the number of nodes and tree depth decreases as alpha
+# increases.
+clfs = clfs[:-1]
+ccp_alphas = ccp_alphas[:-1]
+
+node_counts = [clf.tree_.node_count for clf in clfs]
+depth = [clf.tree_.max_depth for clf in clfs]
+fig, ax = plt.subplots(1, 2)
+ax[0].plot(ccp_alphas, node_counts, drawstyle="steps-post")
+ax[0].set_xlabel("alpha")
+ax[0].set_ylabel("number of nodes")
+ax[0].set_title("Number of nodes vs alpha")
+ax[1].plot(ccp_alphas, depth, drawstyle="steps-post")
+ax[1].set_xlabel("alpha")
+ax[1].set_ylabel("depth of tree")
+ax[1].set_title("Depth vs alpha")
+plt.show()
+
+###############################################################################
+# Accuracy vs alpha for training and testing sets
+# ----------------------------------------------------
+# When ``ccp_alpha`` is set to zero, the decision tree overfits, leading to
+# a 100% training accuracy and 88% testing accuracy. As alpha increases, more
```

Comment:
At the beginning of this example, I was thinking "_here we only show the effect of `cpp_alpha` on regularizing the trees, whereas the other parameters also regularize the tree, although in a slightly different way. Should I suggest doing that in this example? Nah, this example is supposed to focus on `cpp_apha`_".

Then here, you say if `cpp_alpha` is zero, the decision tree overfits; but it only does that if the other parameters are left as defaults on this dataset. I think this part may be a bit misleading. I'd be happy if the comment clarifies that, and I'd be happier if we actually compare the effect of `cpp_alpha`, with let say, `max_depth` and/or another parameter.

Reply:
and maybe here?

Reply:
> but it only does that if the other parameters are left as defaults on this dataset.

Changing the other default parameters may still overfit. I updated this to:

> When ``ccp_alpha`` is set to zero and keeping the other default parameters of `DecisionTreeClassifier`, the tree overfits, leading to a 100% training accuracy and 88% testing accuracy.